File: Data Structures and Problem Solving Using Java 4th/TestSort.java
// Test all the sorting routines in package DataStructures

import weiss.util.Random;

public class TestSort
{
    private static final int NUM_ITEMS = 1000;
    private static int theSeed = 1;
    
    /**
     * Randomly rearrange an array.
     * The random numbers used depend on the time and day.
     * @param a the array.
     */
    public static final <AnyType> void permute( AnyType [ ] a )
    {
        Random r = new Random( );

        for( int j = 1; j < a.length; j++ )
            Sort.swapReferences( a, j, r.nextInt( 0, j ) );
    }

    public static void checkSort( Integer [ ] a )
    {
        for( int i = 0; i < a.length; i++ )
            if( a[ i ] != i )
                System.out.println( "Error at " + i );
        System.out.println( "Finished checksort" );
    }


    public static void main( String [ ] args )
    {
        Integer [ ] a = new Integer[ NUM_ITEMS ];
        for( int i = 0; i < a.length; i++ )
            a[ i ] = i;

        for( theSeed = 0; theSeed < 20; theSeed++ )
        {
            permute( a );
            Sort.insertionSort( a );
            checkSort( a );

            permute( a );
            Sort.heapsort( a );
            checkSort( a );

            permute( a );
            Sort.shellsort( a );
            checkSort( a );

            permute( a );
            Sort.mergeSort( a );
            checkSort( a );

            permute( a );
            Sort.quicksort( a );
            checkSort( a );

            permute( a );
            Sort.quickSelect( a, NUM_ITEMS / 2 );
            System.out.println( a[ NUM_ITEMS / 2 - 1 ] + " is " + NUM_ITEMS / 2 +"th smallest" );
        }
    }
}

File: Data Structures and Problem Solving Using Cpp 2nd/TestSort.cpp


#ifdef USE_DOT_H
    #include <iostream.h>
#else
    #include <iostream>
    #if !defined( __BORLANDC__ ) || __BORLANDC__ >= 0x0530
        using namespace std;      // Borland 5.0 has a bug
    #endif
#endif

#include "Sort.h"
#include "Permute.cpp"

#ifndef SAFE_STL
    #include <vector>
    using namespace std;
#else
    #include "vector.h"
    #include "StartConv.h"
#endif


void checkSort( const vector<int> & a )
{
    for( int i = 0; i < a.size( ); i++ )
        if( a[ i ] != i+1 )
            cout << "Error at " << i << endl;
    cout << "Finished checksort" << endl;
}


int main( )
{
    const int NUM_ITEMS = 1000;

    vector<int> a( NUM_ITEMS );

    for( int theSeed = 0; theSeed < 20; theSeed++ )
    {
        permute( a );
        insertionSort( a );
        checkSort( a );

        permute( a );
        shellsort( a );
        checkSort( a );

        permute( a );
        mergeSort( a );
        checkSort( a );

        permute( a );
        quicksort( a );
        checkSort( a );

        permute( a );
        largeObjectSort( a );
        checkSort( a );

        permute( a );
        quickSelect( a, NUM_ITEMS / 2 );
        cout << a[ NUM_ITEMS / 2 - 1 ] << " " << NUM_ITEMS / 2 << endl;
    }

    return 0;
}

#ifdef SAFE_STL
    #include "EndConv.h"
#endif

